overlord/snapstate/handlers.go

```diff
@@ -503,24 +503,45 @@ func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq map[stri
 		}
 	}
 
+	// If transactional, use a single lane for all tasks, so when
+	// one fails the changes for all affected snaps will be
+	// undone. Otherwise, have different lanes per snap so
+	// failures only affect the culprit snap.
+	var transactionLane int
+	if flags.Transactional {
+		transactionLane = st.NewLane()
+	}
+
 	chg := t.Change()
 	// add all required snaps, no ordering, this will be done in the
 	// auto-connect task handler
 	for _, ts := range tss {
-		ts.JoinLane(st.NewLane())
+		if flags.Transactional {
+			ts.JoinLane(transactionLane)
+		} else {
+			ts.JoinLane(st.NewLane())
+		}
```

Comment:
Suggestion, this and the below conditionals would not be needed with a helper lambda like this:
```
var joinLane func(ts *Task)

if !flags.Transactional {
 joinLane = func(ts *Task) {
    ts.JoinLane(st.NewLane())
 }
} else {
  transactionLane = st.NewLane()
  joinLane = func(ts *Task) {
    ts.JoinLane(transactionLane)
 }
}
```

and then just use `joinLane(ts)` everywhere.

Reply:
Nice one, I've done this now.